### DISCUSSION TITLE
Save audio from I2S mic to .wav file on SD card

### DISCUSSION THREAD

**Discussion by rasmushauschild** [pschatzmann/arduino-audio-tools#440] [Q&A] [answered]
Hi!
As the title alludes to, I would like to save the audio recorded on an I2S mic to a .wav file on an SD card. 
I'm able to read files from the SD card just fine, and I can print the I2S data from the mic to the serial monitor using a `CsvStream` as the `StreamCopy ` output. However, when I change the `StreamCopy ` output to a file on the SD card, the file itself is created, but it's empty.
I imagine I must be missing something basic, but I'm having a hard time figuring out what.

This is my code:

```
#include <SPI.h>
#include "Arduino.h"
#include "AudioTools.h"
#include "SD.h"

#define SD_CS     5
#define SD_MOSI   33
#define SD_MISO   1
#define SD_SCK    9

SPIClass SPISD(HSPI); // Creates SPI object using HSPI

unsigned long runtime;

I2SStream i2s; // Access I2S as a stream

File audioFile; //audio file object
EncodedAudioStream out(&audioFile, new WAVEncoder()); // Access audioFile as a stream

StreamCopy copier(out, i2s); // copies sound from i2s to audio file stream


// Arduino Setup
void setup(void) {  

  Serial.begin(115200);

  // change to Warning to improve the quality
  AudioLogger::instance().begin(Serial, AudioLogger::Info); 

  Serial.println("SERIAL STARTED");

  SPISD.begin(SD_SCK, SD_MISO, SD_MOSI); // Begin SPI used for SD card

  if(!SD.begin(SD_CS, SPISD)) {
    Serial.println("SPI FAILED");
  }

  Serial.println(audioFile.readString());

  // start I2S in
  Serial.println("starting I2S...");

  auto cfg = i2s.defaultConfig(RX_MODE);
  cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
  cfg.bits_per_sample = 32;
  cfg.channels = 1;
  cfg.sample_rate = 16000;
  cfg.pin_ws = 14;
  cfg.pin_bck = 12;
  cfg.pin_data = 38;
  i2s.begin(cfg);

  auto cfg_out = out.defaultConfig();
  cfg_out.bits_per_sample = 32;
  cfg_out.channels = 1;
  cfg_out.sample_rate = 16000;
  out.begin(cfg_out);

  Serial.println("I2S started...");

  audioFile = SD.open("/AudioFile01.wav", FILE_WRITE);
  Serial.println("RECORDING");
}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();
  runtime = millis();

  if (runtime > 10000) {
    audioFile.close();
    Serial.println("FILE CLOSED");
  }

}
```

Any help is greatly appreciated! 🙏

**Comment by pschatzmann** [comment]
Not sure what exactly the issue is, but I think you should
- call copier.copy only when the file is open
- try to decrease the complexity to identify the problem by copying to the file directly.

I am not sure if the approach to set up the logic before the file really exists does work. In my sketches I used to do this setup only after I had an open file, by calling the corresponding begin methods: see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-audiokit-sd-audiokit/streams-audiokit-sd-audiokit.ino

**Comment by rasmushauschild** [comment, marked as answer]
Thanks for the input!
I tried to implement the changes you proposed, however, I still can't seem to get it to write anything in the file that is being created...

This is the updated sketch:
```
#include <TFT_eSPI.h>
#include <SPI.h>
#include "Arduino.h"
#include "AudioTools.h"
#include "SD.h"

#define SD_CS     5
#define SD_MOSI   33
#define SD_MISO   1
#define SD_SCK    9



SPIClass SPISD(HSPI); // Creates SPI object using the HSPI interface

bool recording = false;
unsigned long runtime;


I2SStream i2s; // Access I2S as a stream

File audioFile; //empty audio file object
EncodedAudioStream out(&audioFile, new WAVEncoder()); // Access audioFile as a stream

StreamCopy copier(out, i2s); // copies sound from i2s to audio file stream



// Arduino Setup
void setup(void) {  
  Serial.begin(115200);

  AudioLogger::instance().begin(Serial, AudioLogger::Debug); // AudioLogger set to Debug mode

  Serial.println("SERIAL STARTED");

  SPISD.begin(SD_SCK, SD_MISO, SD_MOSI); // Begin SPI used for SD card

  if(!SD.begin(SD_CS, SPISD)) {
    Serial.println("SPI FAILED");
  }

  // configure I2S stream
  Serial.println("starting I2S...");

  auto cfg = i2s.defaultConfig(RX_MODE);
  cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
  cfg.bits_per_sample = 32;
  cfg.channels = 2;
  cfg.sample_rate = 16000;
  cfg.pin_ws = 14;
  cfg.pin_bck = 12;
  cfg.pin_data = 38;
  i2s.begin(cfg);

  // Configure out stream
  auto cfg_out = out.defaultConfig();
  cfg_out.bits_per_sample = 32;
  cfg_out.channels = 2;
  cfg_out.sample_rate = 16000;
  out.begin(cfg_out);
}

// Arduino loop
void loop() {
  runtime = millis();

  if (!recording) {
    audioFile = SD.open("/AudioFile01.wav", FILE_WRITE);  // Open a new file on the SD card for writing to
    copier.begin(out, i2s);  // Configure copier input and output
    Serial.println("RECORDING");
    recording = true;  // Makes sure this only runs once
  }

  if (runtime > 5000) {
    audioFile.close(); // Closes the file after 5 seconds
    // Serial.println("FILE CLOSED");
  } else copier.copy(); // Copy sound from I2S to the audio file

}
```

I set AudioLogger to Debug, and this is what I get:

```
SERIAL STARTED
starting I2S...
[D] I2SStream.h : 94 - void audio_tools::I2SStream::begin(audio_tools::I2SConfig)        
[D] I2SESP32.h : 42 - bool audio_tools::I2SBase::begin(audio_tools::I2SConfig)
[D] I2SESP32.h : 138 - bool audio_tools::I2SBase::begin(audio_tools::I2SConfig, int, int)
[I] I2SConfig.h : 98 - rx/tx mode: RX_MODE
[I] I2SConfig.h : 99 - port_no: 0
[I] I2SConfig.h : 100 - is_master: Master
[I] I2SConfig.h : 101 - sample rate: 16000
[I] I2SConfig.h : 102 - bits per sample: 32
[I] I2SConfig.h : 103 - number of channels: 2
[I] I2SConfig.h : 104 - i2s_format: I2S_STD_FORMAT
[I] I2SConfig.h : 106 - auto_clear:0
[I] I2SConfig.h : 113 - buffer_count:10
[I] I2SConfig.h : 114 - buffer_size:512
[I] I2SConfig.h : 118 - pin_bck: 12
[I] I2SConfig.h : 119 - pin_ws: 14
[I] I2SConfig.h : 120 - pin_data: 38
[D] I2SESP32.h : 166 - i2s_driver_install
[D] I2SESP32.h : 183 - i2s_set_pin
[D] I2SESP32.h : 194 - i2s_zero_dma_buffer
[D] I2SESP32.h : 198 - begin - started
[D] AudioEncoded.h : 304 - void audio_tools::EncodedAudioStream::begin(audio_tools::AudioBaseInfo)
[I] CodecWAV.h : 446 - is_streamed! because length is 2147418112
[I] AudioCopy.h : 74 - buffer_size=1024
RECORDING
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
[D] AudioCopy.h : 91 - size_t audio_tools::StreamCopyT<T>::copy() [with T = unsigned char; size_t = unsigned int]
```

I'm not sure what it's supposed to say for a case like this, but I can't detect any errors myself - can you tell if something is wrong from the debug log?

Thank you for your help! I appreciate it 🙏

**Reply by pschatzmann** [reply]
Yes, I would have expected to see some copy info log entries to confirm that something is copied. Maybe the file returns availableForWrite() with 0 which would cause this...
I suggest that you try to call copier.setCheckAvailableForWrite(false) to deactivate this check...

**Reply by rasmushauschild** [reply]
Amazing! That seems to be what caused the issue - I can now record to the SD card just as expected.
Any idea why `availableForWrite()` is returning false in this case?

**Reply by ksaye** [reply]
@Rasmushauschild, can you share your final and were you able to use SPIFFS?  I am now under the impression that SPIFFS is too slow.

I am trying the following code, and while I can produce a valid wav file, the file is not a 10 second file.  I am using the PCM5102A device.

      #include "SPIFFS.h"
      #include "AudioTools.h"
      
      #define MicPIN_WS GPIO_NUM_33
      #define MicPIN_SCK GPIO_NUM_32
      #define MicPIN_SD GPIO_NUM_36
      
      bool recording = false;
      unsigned long runtime;
      
      I2SStream i2s; // Access I2S as a stream
      File audioFile; //empty audio file object
      EncodedAudioStream out(&audioFile, new WAVEncoder()); // Access audioFile as a stream
      StreamCopy copier(out, i2s); // copies sound from i2s to audio file stream
      
      // Arduino Setup
      void setup(void) {  
        Serial.begin(115200);
      
        AudioLogger::instance().begin(Serial, AudioLogger::Info); // AudioLogger set to Debug mode
      
        Serial.println("SERIAL STARTED");
      
        SPIFFS.begin(true);
      
        // configure I2S stream
        Serial.println("starting I2S...");
      
        auto cfg = i2s.defaultConfig(RX_MODE);
        //cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
        //cfg.i2s_format = I2S_LSB_FORMAT; // or try with 
        //cfg.bits_per_sample = 16;
        cfg.channels = 2;
        cfg.sample_rate = 16000;
        cfg.pin_ws = MicPIN_WS;       //WS
        cfg.pin_bck = MicPIN_SCK;     //SCK
        cfg.pin_data = MicPIN_SD;     //SD
        i2s.begin(cfg);
      
        // Configure out stream
        auto cfg_out = out.defaultConfig();
        //cfg_out.bits_per_sample = 32;
        cfg_out.channels = 2;
        cfg_out.sample_rate = 16000;
        out.begin(cfg_out);
        copier.setCheckAvailableForWrite(false);
      }
      
      // Arduino loop
      void loop() {
        runtime = millis();
      
        if (!recording) {
          audioFile = SPIFFS.open("/AudioFile01.wav", FILE_WRITE);  // Open a new file on the SD card for writing to
          copier.begin(out, i2s);  // Configure copier input and output
          Serial.println("RECORDING");
          recording = true;  // Makes sure this only runs once
        }
      
        if (runtime > 10000) {
          audioFile.close(); // Closes the file after 5 seconds
          Serial.println("FILE CLOSED");
        } else copier.copy(); // Copy sound from I2S to the audio file
      
      }

One item of interest is that the Bit Rate, when viewed through Windows, is 512 kbps.

![image](https://user-images.githubusercontent.com/15837044/223561768-3663cd17-e437-4cdd-a24e-ea5f1d6e9cb5.png)

If I specify the Bit Rate to 32, as you have done, I do see a 1024 kbps:

![image](https://user-images.githubusercontent.com/15837044/223564209-ac6332cc-d811-4fd8-ba40-1bffedb64989.png)

the log file throws no errors:

      15:56:25.062 -> SERIAL STARTED
      15:56:25.186 -> starting I2S...
      15:56:25.186 -> [I] I2SConfig.h : 104 - rx/tx mode: RX_MODE
      15:56:25.186 -> [I] I2SConfig.h : 105 - port_no: 0
      15:56:25.186 -> [I] I2SConfig.h : 106 - is_master: Master
      15:56:25.186 -> [I] I2SConfig.h : 107 - sample rate: 16000
      15:56:25.186 -> [I] I2SConfig.h : 108 - bits per sample: 16
      15:56:25.186 -> [I] I2SConfig.h : 109 - number of channels: 2
      15:56:25.218 -> [I] I2SConfig.h : 110 - i2s_format: I2S_STD_FORMAT
      15:56:25.218 -> [I] I2SConfig.h : 112 - auto_clear:0
      15:56:25.218 -> [I] I2SConfig.h : 119 - buffer_count:10
      15:56:25.218 -> [I] I2SConfig.h : 120 - buffer_size:512
      15:56:25.218 -> [I] I2SConfig.h : 124 - pin_bck: 32
      15:56:25.218 -> [I] I2SConfig.h : 126 - pin_ws: 33
      15:56:25.218 -> [I] I2SConfig.h : 128 - pin_data: 36
      15:56:25.218 -> [I] CodecWAV.h : 452 - sample_rate: 16000
      15:56:25.218 -> [I] CodecWAV.h : 453 - channels: 2
      15:56:25.263 -> [I] CodecWAV.h : 457 - is_streamed! because length is 2147418112
      15:56:26.565 -> [I] AudioCopy.h : 79 - buffer_size=1024
      15:56:26.565 -> RECORDING
      15:56:26.565 -> [I] CodecWAV.h : 501 - Writing Header
      15:56:29.894 -> [I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
      15:56:33.382 -> [I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
      15:56:37.187 -> [I] AudioCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
      15:56:38.967 -> FILE CLOSED

[16 bit rate AudioFile01.zip](https://github.com/pschatzmann/arduino-audio-tools/files/10914198/AudioFile01.zip)
[32 bit rate AudioFile01 (1).zip](https://github.com/pschatzmann/arduino-audio-tools/files/10914283/AudioFile01.1.zip)
